light_curves/code_src/ztf_functions.py

```diff
@@ -10,316 +11,91 @@
-            indexes_to_keep.extend(singleband_object.index)
-        else:
-            npoints = singleband_object["mag"].str.len()
-            npointsmax_object = singleband_object.loc[npoints == npoints.max()]
-            # this may still have more than one light curve if they happen to have the same number
-            # of datapoints (e.g., Yang sample sample_table[7], band 'zr').
-            # arbitrarily pick the one with the min oid.
-            # depending on your science, you may want (e.g.,) the largest timespan instead
-            if len(npointsmax_object.index) == 1:
-                indexes_to_keep.extend(npointsmax_object.index)
-            else:
-                minoid = npointsmax_object.oid.min()
-                indexes_to_keep.extend(npointsmax_object.loc[npointsmax_object.oid == minoid].index)
-    ztf_df = ztf_df.loc[sorted(indexes_to_keep)]
-
-    # store "hmjd" as "time".
-    # note that other light curves in this notebook will have "time" as MJD instead of HMJD.
-    # if your science depends on precise times, this will need to be corrected.
-    ztf_df = ztf_df.rename(columns={"hmjd": "time"})
-
-    # "explode" the data structure into one row per light curve point and set the correct dtypes
-    # note that this operation may require 6+ times the RAM used for ztf_df up to now
-    ztf_df = ztf_df.explode(["time", "mag", "magerr", "catflags"], ignore_index=True)
-    ztf_df = ztf_df.astype({"time": "float", "mag": "float", "magerr": "float", "catflags": "int"})
-
-    # remove data flagged as bad
-    ztf_df = ztf_df.loc[ztf_df["catflags"] < 32768, :]
-
-    # calc flux [https://arxiv.org/pdf/1902.01872.pdf zeropoint corrections already applied]
-    mag = ztf_df["mag"].to_numpy()
-    magerr = ztf_df["magerr"].to_numpy()
-    fluxupper = ((mag - magerr) * u.ABmag).to_value('mJy')
-    fluxlower = ((mag + magerr) * u.ABmag).to_value('mJy')
-    ztf_df["flux"] = (mag * u.ABmag).to_value('mJy')
-    ztf_df["err"] = (fluxupper - fluxlower) / 2
+    # 1) fetch the raw arrays 
+    raw = hats_get_lightcurves(
+        sample_table,
+        object_catalog_path=UPath('s3://irsa-fornax-testdata/ZTF/dr23/objects/hats'),
+        object_margin_cache=None,
+        lightcurve_catalog_path=UPath('s3://irsa-fornax-testdata/ZTF/dr23/lc/hats'),
+        lightcurve_margin_cache=None,
+        object_columns=["oid", "ra", "dec"],
+        light_curve_columns=["objectid", "hmjd", "filterid", "mag", "magerr", "catflags"],
+        id_col='objectid',
+        time_col='hmjd',
+        flux_col='mag',    # this becomes raw.data['flux']
+        err_col='magerr',  # this becomes raw.data['err']
+        radius=radius,
+        filter_id_to_name=None,
+    )
+    # 2) turn the MultiIndex into columns
+    df = raw.data.reset_index()
+
+    # 3) rename columns
+    df = df.rename(columns={"hmjd": "time"})
+    #these are called flux and err in the generic hats_get_lightcurves, but are really magnitudes for ztf
```

Comment:
I would avoid having `hats_get_lightcurves` change the column names for this very reason.